pkg/skaffold/build/custom/custom_test.go

```diff
@@ -125,8 +130,98 @@ func TestRetrieveCmd(t *testing.T) {
 	}
 }
 
+func TestGracefulBuildCancel(t *testing.T) {
+	if runtime.GOOS == "windows" {
+		t.Skip("graceful cancel doesn't work on windows")
+	}
+
+	tests := []struct {
+		artifact    *latest.Artifact
+		testEnv     []string
+		shouldErr   bool
+		description string
+	}{
+		{
+			description: "terminate gracefully and exit 0",
+			artifact: &latest.Artifact{
+				ArtifactType: latest.ArtifactType{
+					CustomArtifact: &latest.CustomArtifact{
+						BuildCommand: fmt.Sprintf("%s --test.run=TestHelperGracefulBuildCancel", os.Args[0]),
+					},
+				},
+			},
+			testEnv:   []string{"GO_STARTED_FROM_TEST=1"},
+			shouldErr: false,
+		},
+		{
+			description: "terminate gracefully and kill process",
+			artifact: &latest.Artifact{
+				ArtifactType: latest.ArtifactType{
+					CustomArtifact: &latest.CustomArtifact{
+						BuildCommand: fmt.Sprintf("%s --test.run=TestHelperGracefulBuildKill", os.Args[0]),
+					},
+				},
+			},
+			testEnv:   []string{"GO_STARTED_FROM_TEST=1"},
+			shouldErr: true,
+		},
+	}
+
+	for _, test := range tests {
+		testutil.Run(t, test.description, func(t *testutil.T) {
+
+			builder := NewArtifactBuilder(false, test.testEnv)
+
+			ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+
+			err := builder.Build(ctx, os.Stdout, test.artifact, "image:tag")
+			t.CheckError(test.shouldErr, err)
+
+			cancel()
+		})
+	}
+}
+
+func TestHelperGracefulBuildCancel(t *testing.T) {
+	if os.Getenv("GO_STARTED_FROM_TEST") != "1" {
+		return // skip test if not started from a test case
+	}
+
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, os.Interrupt)
+
+	go func() {
+		<-signals
+		fmt.Println("Got interrupt")
+	}()
+
+	fmt.Println("Sleeping 25 seconds")
```

Comment:
i am little bit concerned with these tests adding sleeps. Can you just add a bash command like `sleep 5` in the build script?

Also, run it multiple times (100) to make sure its reliable.

Reply:
Imo 100 repeats are a bit much for a unit test. Also inlining a bash script required some extra changes in code.

Reply:
Also the tests require the interrupt signal to be trapped. Otherwise they will always error.

Reply:
sorry, i meant running it 100 times via `go run -n -test.run <>`. I will do it myself before reviewing.